feat: Proving for the playground

Dominik1999 opened the pull request:
Closes #19 

This PR enables proving to the playground. 

There are some open questions regarding the UX and the proving step code itself:

- UX
- - At the moment, for the user running and proving a program looks equal. The output is almost the same. 
- - We should add some info that proving might take a while (see Game of Life example), maybe even a pop-up that prints something during the proof
- - There is no way to show the proof atm

- Proving step
- - We don't expose in Miden v0.3 any info about the trace length or cycles or even time to prove in `ProgramOutputs` or `ExecutionTrace`, so we can't get the numbers (maybe I am mistaken, but I didn't find it)
- - I couldn't find a way to pass the proof from Rust to Typescript (frontend) via wasm-bindgen. There seems to be a way to return complex structures from Rust to Typescript (frontend) but I didn't understand exactly how

grjte commented:
Since this will need to be rebased and updated after the repo structure PR is changed, let's keep it as a draft until then. I'm going to put it into draft mode and remove my review request. Once it's been rebased, please mark it as "ready for review" and then re-request review

grjte commented:
> There are some open questions regarding the UX and the proving step code itself:
> 
>     * UX
> 
>     * * At the moment, for the user running and proving a program looks equal. The output is almost the same.

I think this is addressed by discussions below. As long as we have a way to provide/show the proof then it's fine for the output to be the same (imo). Is there something else you were thinking about here?

>     * * We should add some info that proving might take a while (see Game of Life example), maybe even a pop-up that prints something during the proof

We can use react-hot-toast to do this very easily: https://react-hot-toast.com/

> 
>     * * There is no way to show the proof atm

I think we should create a new issue for this, but reference #19 inside that issue, as well as the other related open questions/issues you've noted

>     * Proving step
> 
>     * * We don't expose in Miden v0.3 any info about the trace length or cycles or even time to prove in `ProgramOutputs` or `ExecutionTrace`, so we can't get the numbers (maybe I am mistaken, but I didn't find it)

I think this is part of a larger conversation in the VM, and related to changes for debugging improvements as well (https://github.com/0xPolygonMiden/miden-vm/issues/580). We should bring it up on the team sync call, but maybe this should be separated out from this set of playground chanages and become its own working group that looks at the 2 things together.

>     * * I couldn't find a way to pass the proof from Rust to Typescript (frontend) via wasm-bindgen. There seems to be a way to return complex structures from Rust to Typescript (frontend) but I didn't understand exactly how

This seems to be related to the comment above about having a way to show the proof, so it seems to me that we should just create a new issue for showing the proof.

Dominik1999 commented:
> Since this will need to be rebased and updated after the repo structure PR is changed, let's keep it as a draft until then. I'm going to put it into draft mode and remove my review request. Once it's been rebased, please mark it as "ready for review" and then re-request review

Thanks, I just rebased.

Dominik1999 commented:
> > ```
> > * * I couldn't find a way to pass the proof from Rust to Typescript (frontend) via wasm-bindgen. There seems to be a way to return complex structures from Rust to Typescript (frontend) but I didn't understand exactly how
> > ```
> 
> This seems to be related to the comment above about having a way to show the proof, so it seems to me that we should just create a new issue for showing the proof.

Here is a new issue about getting and showing the proof. We can tackle this in the next milestone. https://github.com/0xPolygonMiden/examples/issues/52

grjte commented:
> Here is a new issue about getting and showing the proof. We can tackle this in the next milestone. #52

Sounds good! I think we should also create an issue about adding toast messages and put that in the "nice to have" for this milestone. wdyt?

grjte commented on playground/miden-wasm/src/lib.rs:
```diff
@@ -63,16 +63,61 @@ pub fn program(asm: &str, inputs_frontend: &str, output_count: u16) -> Vec<u64>
+    let mut overflow_addrs = output.overflow_addrs().to_vec();
+    stack_outputs.append(&mut overflow_addrs);
+    
+    // TODO: We must return trace_info as well, but we can not handle that yet in Miden v0.3.
+    // TODO: We must return the proof as well. But we can not handle that yet in the frontend.
```
Let's just make sure these are both captured in an issue (I think they are, but can you confirm?)

Dominik1999 replied:
Yes, it's tracked in https://github.com/0xPolygonMiden/examples/issues/52

grjte commented on playground/miden-wasm/src/lib.rs:
```diff
@@ -63,16 +63,61 @@ pub fn program(asm: &str, inputs_frontend: &str, output_count: u16) -> Vec<u64>
+    let mut stack_init = <Vec<u64>>::new();
+    let mut advice_tape = <Vec<u64>>::new();
+
+    // TODO: We must catch all cases of inputs. But maybe in the frontend already.
```
I don't understand this. Can you rephrase and also make sure it's tracked by an issue?

Dominik1999 replied:
This is tracked here - and the next PR that I will send to you https://github.com/0xPolygonMiden/examples/issues/18

I will rephrase

grjte commented on playground/miden-wasm/src/lib.rs:
```diff
@@ -63,16 +63,61 @@ pub fn program(asm: &str, inputs_frontend: &str, output_count: u16) -> Vec<u64>
+    
+    // We convert the outputs into one Vector because we can't return tuples in wasm-bindgen.
+    // So the first n elements are the stack_outputs and the last n elements are the overflow_addrs.
+    // TODO: We can somehow refactor to return structures in wasm-bindgen. But I didn't find a way yet.
+    // There might be a way to return the ProgramOutputs struct even.
```
We can leave this for another PR. Let's just make sure it's tracked in an  issue

Dominik1999 replied:
this is also tracked in https://github.com/0xPolygonMiden/examples/issues/52

grjte commented on playground/miden-wasm/src/lib.rs:
```diff
@@ -63,16 +63,61 @@ pub fn program(asm: &str, inputs_frontend: &str, output_count: u16) -> Vec<u64>
+        let inputs_des: InputFile = serde_json::from_str(&inputs_frontend)
+        .map_err(|err| format!("Failed to deserialize input data - {}", err)).unwrap();
+        
+        stack_init = get_stack_init(&inputs_des);
+        advice_tape = get_advice_tape(&inputs_des);
+    } 
+
+    let inputs = get_program_inputs(&stack_init, &advice_tape); 
+
+    // default (96 bits of security)
+    let proof_options = ProofOptions::with_96_bit_security();
+
+    let (output, _proof) = miden::prove(
+        &program,
+        &inputs,
+        &proof_options,
+    )
+    .map_err(|err| format!("Failed to generate exection trace = {:?}, and advice_tape = {:?}", err, advice_tape)).unwrap();
+    
+    // We convert the outputs into one Vector because we can't return tuples in wasm-bindgen.
+    // So the first n elements are the stack_outputs and the last n elements are the overflow_addrs.
+    // TODO: We can somehow refactor to return structures in wasm-bindgen. But I didn't find a way yet.
+    // There might be a way to return the ProgramOutputs struct even.
+    let mut stack_outputs = output.stack_outputs(output_count as usize).to_vec();
+    let mut overflow_addrs = output.overflow_addrs().to_vec();
+    stack_outputs.append(&mut overflow_addrs);
+    
+    // TODO: We must return trace_info as well, but we can not handle that yet in Miden v0.3.
+    // TODO: We must return the proof as well. But we can not handle that yet in the frontend.
+
+    stack_outputs
+
+}
+
 /// Basic test for the program
 #[wasm_bindgen_test]
-fn run_program() {
-    let output = program(
+fn test_run_program() {
+    let output = run_program(
```
Does it make sense to add a basic test for proving too?

Dominik1999 replied:
Hmm indeed, we should add that.

Dominik1999 replied:
added a test

grjte reviewed (changes requested):
Looks great, thank you! I left a few very minor comments. Also, I want to be sure all inline TODOs are captured in issues. I think many/most are, but just to check

grjte commented on playground/miden-wasm/src/lib.rs:
```diff
@@ -63,16 +63,62 @@ pub fn program(asm: &str, inputs_frontend: &str, output_count: u16) -> Vec<u64>
+    let mut stack_init = <Vec<u64>>::new();
+    let mut advice_tape = <Vec<u64>>::new();
+
+    // TODO: We must catch all cases of inputs. Input can be not a valid json,
+    // input can be empty, stack_init can be empty, advice_tape can be empty. And all combinations.
```
nit: I see - you're talking about cases of input errors. I would probably have said "all cases of input errors", but the message is now clear, and no need to update, especially since it'll be tackled in the next PR.

Dominik1999 replied:
ah indeed. I like it those nits

grjte reviewed (approved):
Approved! Thank you!
